pmoves/docs/context/copy_of_sanl_—_spiral_atlas_number_learner.py

```
@@ -0,0 +1,434 @@
+
+def unit_penalty(head):
+    """
+    head: [B, H, 2]
+    Encourage |χ| ≈ 1.
+    """
+    mag = torch.norm(head, dim=-1)
+    return ((mag - 1.0)**2).mean()
+
+def ortho_penalty_over_group(model, atlas, p: int, sample: int = 2048):
+    """
+    Encourage heads to behave like orthogonal characters on the group (Z/pZ)^×.
```

Ruff (0.14.8)

Docstring contains ambiguous `×` (MULTIPLICATION SIGN). Did you mean `x` (LATIN SMALL LETTER X)?

(RUF002)